python/lsst/ctrl/mpexec/quantumGraphExecutor.py

```diff
@@ -0,0 +1,185 @@
+# This file is part of ctrl_mpexec.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ['QuantumGraphExecutor']
+
+# -------------------------------
+#  Imports of standard modules --
+# -------------------------------
+from abc import ABC, abstractmethod
+import logging
+
+# -----------------------------
+#  Imports for other modules --
+# -----------------------------
+
+_LOG = logging.getLogger(__name__.partition(".")[2])
+
+
+class QuantumGraphExecutor(ABC):
+    """Class with partial implementation of QunatumGraph execution.
+
+    Any specific execution model is implemented in sub-class by overriding
+    the `executeQuanta` method.
+    """
+    def __init__(self):
+        pass
+
+    def execute(self, graph, butler, taskFactory, registerDatasetTypes=False,
+                saveInitOutputs=True, updateOutputCollection=True, initOnly=False):
+        """Execute whole graph.
+
+        Parameters
+        ----------
+        graph : `~lsst.pip.base.QuantumGraph`
+            Execution graph.
+        butler : `~lsst.daf.butler.Butler`
+            Data butler instance
+        taskFactory : `~lsst.pipe.base.TaskFactory`
+            Task factory.
+        registerDatasetTypes : `bool`, optional
+            If ``True`` then register dataset types in registry, otherwise
+            they must be already registered.
+        saveInitOutputs : `bool`, optional
+            If ``True`` (default) then save task "init outputs" to butler.
+        updateOutputCollection : `bool`, optional
+            If ``True`` (default) then copy all inputs to output collection.
+        initOnly : `bool`, optional
+            if ``True`` then only initialization part is run, but not any
+            task.
```

Comment:
I was thinking about making `executeQuanta()` a no-op method rather than abstract, then you could instantiate this class and use it on a head node directly but I agree it looks awkward. Let me try to refactor this into a separate class (suggestions for a good class name are welcome).